src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,215 @@
+            } else {
+                node = head;
+                for (int i = 0; i < size; i++) {
+                    if (i == index) {
+                        node = new Node<>(node.prev,value,node);
+                        node.prev.next = node;
+                        break;
+                    }
+                    node = node.next;
+                }
+                size++;
+            }
+
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
+        Node<T> node = head;
+        if (index > 0 || index > (size - 1)) {
+            for (int i = 0; i < size; i++) {
+                if (i == index) {
+                    return node.value;
+                } else {
+                    if (node.equals(tail)) {
+                        throw new IndexOutOfBoundsException("size; " + size + "index: " + index);
+                    }
+                    node = node.next;
+                }
+            }
+        } else if (index == 0) {
+            return head.value;
+        } else {
+            throw new IndexOutOfBoundsException("size; " + size + "index: " + index);
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = head;
+        T oldValue = null;
+        if ((size - 1) > index) {
+            for (int i = 0; i < size; i++) {
+                if (i == index) {
+                    oldValue = node.value;
+                    node.value = value;
+                    break;
+                }
+                node = node.next;
+            }
+            return oldValue;
+        } else {
+            throw new IndexOutOfBoundsException("size; " + size + "index: " + index);
+        }
+
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = head;
+        if (index > 0 && index < (size)) {
+            for (int i = 0; i < size; i++) {
+                if (i == index) {
+                    if (node.next != null) {
+                        node.next.prev = node.prev;
+                    }
+                    node.prev.next = node.next;
+                    size--;
+                    return node.value;
+                } else {
+                    node = node.next;
+                }
+            }
+        } else if (index == 0) {
+            if (size != 1) {
+                node.next.prev = null;
+                head = node.next;
+                size--;
+                return node.value;
+            } else {
+                head = null;
+                tail = null;
+                size--;
+                return node.value;
+            }
+
+        }
+        {
+            throw new IndexOutOfBoundsException("size; " + size + "index: " + index);
+        }
     }
```

Comment:
In the `remove(int index)` method, you should create a separate method `unlink(Node node)` to remove all the links when you remove a Node. This logic will be used in 2 methods: when we remove from the list by index and by value. So let's create a method that will take a Node that needs to be removed and unlink it. As a result, we will call it from both methods when Node is found.